homeassistant/bootstrap.py

```diff
@@ -98,6 +110,22 @@ def _handle_requirements(hass: core.HomeAssistant, component,
     return True
 
 
+def _async_init_setup_data(hass: core.HomeAssistant, components: List[str]):
+    """Initializing async setup construct."""
+    setup_events = hass.data.get(DATA_SETUP_EVENTS)
+    if setup_events is None:
+        setup_events = hass.data[DATA_SETUP_EVENTS] = {}
+
+    for domain in components:
+        if domain in setup_events:
+            continue
+
+        setup_events[domain] = {}
```

Comment:
I just realized that we can make it even more beautiful by just using what is available: `asyncio.Future`.

- Have `hass.async_add_job` return the created task. 
- Store that task in the dictionary

We can then do:

```python
setup_events[domain].done()
setup_events[domain].result()
yield from setup_events[domain]
```

Yielding multiple times from a finished task is not a problem, if it is finished it will return the already available result.

So then setup can become:

```python
if domain not in setup_events:
    setup_events[domain] = hass.async_add_job(async_setup_component(…))

# minor optimization: do not do extra yield if result is already available
if setup_events[domain].done():
    result = setup_events[domain].result()
else:
    result = yield from setup_events[domain]

return result
```